### PR TITLE
Search Page

### PR DESCRIPTION
**Pull request by ikesau**
_Last update: Sept 6 by @mlbrgl_ 

[Staging demo](http://staging-site-search-page/)
_outdated_ ~~[Ptolemy demo](https://ptolemy-owid.netlify.app/)~~

[Figma reference](https://www.figma.com/file/36r4ZDrt0c4YdOGVduSjQm/%5BProject%5D-Search?type=design&node-id=196-18241&mode=design&t=0RuGlGybKgbTJ70a-0)

This PR has two major components:
1. `SearchPanel` (an [InstantSearch](https://www.algolia.com/doc/api-reference/widgets/instantsearch/react/) integration)
2. `Autocomplete` (an [Autocomplete](https://www.algolia.com/doc/ui-libraries/autocomplete/introduction/getting-started/) integration)

InstantSearch is a React-first, hook based solution.

Autocomplete _isn't_ a react-first library, so there's some funky stuff going on to update data and render things.

There's quite a few band-aids. If any of the code looks slightly confusing/weird, please flag it and I'll add a comment explaining it. 

### Before launch (v1)
- [x] Remove old search JS and CSS
- [x] Analytics
- [x] confirm autocomplete fuzzy matching behaviour (mockups have string match, current implementation is fuzzy matching: is this what we want?)
- [x] Bundlesize measurements (checked by @marcelgerber, mostly due to the addition of Algolia's packages)
- [x] the explorers index configuration is missing in the configureAlgolia script
- [x] console warnings
- [x] Bug bash: critical vs nice to have
- [x] show autocomplete on search page ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1692881570075319))
- [x] critical design QA updates
- [x] missing highlighted term in autocomplete ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1692870632159329))
- [ ] (launch) add explorer indexing to crontab

###  v2
- [ ] faceting
- [ ] add search bar on homepage
- [ ] "no active search" screen ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1692878773626479))
- [ ] use createInsightsMiddleware? https://github.com/owid/owid-grapher/pull/2533#discussion_r1307955755
- [ ] search: show featured images for posts ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693322933084609))
- [ ] placement ~~and color~~ ✅ of content type in autocomplete ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1692871062357549))
- [x] placement of "show all" link ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693213575757329?thread_ts=1692880180.450959&cid=C04N12KT6GY)) 
- [x] search box slightly taller than the other buttons ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1692981064542879))
- [ ] smoother transition when opening the autocomplete ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693305233399299))
- [ ] (autocomplete) reduce "X" 16px ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693314251119739))
- [ ] (search bar) The text seems to be black here but should be #1D3D63 ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693315174650219))
- [ ] (search bar) x icon should be 16px in search bar ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693315174650219))
- [ ] (search bar) search icon is missing ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693315174650219))
- [ ] there should be 24px between the search bar and the content switcher ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693315540232799))
- [ ] (filters) I think it would make sense to extend this line to go full-width (like the navbar) ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693316490294219))
- [ ] (filters) the numbers should be 10px ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693316490294219))
- [ ] (filters) the color of the search box outline and the line should be the same #DBE5F0 ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693316490294219))
- [ ] (r&w) missing excerpts for several pages ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693337203539749))
- [ ] the Data Explorer grey blocks maybe look a bit bare/redundant without anything else there ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693337737394149))
- [ ] (mobile) Since we don’t have the “Search results” title anymore, I think we should have the titles of the sections “R&W”, “Charts” in Blue 50 (like on desk) ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693387026045399))
- [ ] (mobile) The number near each title is missing ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693387026045399))
- [ ] (mobile) The vertical spacing seems off in general.. see this screen for reference ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693387026045399))
- [ ] (mobile) Button should be 40px high ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693388596855999))
- [ ] (mobile) Can we have “Showing 4 of 40 results” in Blue 90 on Mobile? ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693388596855999))
- [ ] (mobile) Again here spacing looks off… Show all is too close to the next section + too much space between R&W block and “Showing 4/40 results” ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693388596855999))
- [ ] (mobile) Data explorers should we placed horizontally ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1693389798335309))
- [ ] (maybe) sticky nav bar reappears when scrolling up ([slack](https://owid.slack.com/archives/C04N12KT6GY/p1692798001675179))

### CODE REVIEW

File: site/search/Autocomplete.tsx
Comment (mlbrgl) [inline site/search/Autocomplete.tsx:192]:
```suggestion
        if (window.location.pathname.startsWith("/search")) return
```

File: site/search/Autocomplete.tsx
Comment (mlbrgl) [inline site/search/Autocomplete.tsx:182]: We'll cross that bridge when we get there but we'll probably have to adapt some of the logic which assumes that there can only be one autocomplete per page, which won't be the case on e.g. the homepage.

File: site/search/Autocomplete.tsx
Comment (mlbrgl) [inline site/search/Autocomplete.tsx:191]: I don't get why using `useLayoutEffect` vs `useEffect` here

**Review by mlbrgl — APPROVED** [review]
It's great that you managed to use so much of the Algolia's API for this.

A solid implementation, good work!

On top of the things you noted, here are a few additional items we might need to be addressed before merging:
- the explorers index configuration is missing in the configureAlgolia script.
- algolia credit
- a couple of warnings in the console

Otherwise a very good step!

File: site/search/SearchPanel.tsx
Comment (ikesau) [reply site/search/SearchPanel.tsx:238]: I agree this part is a bit weird, I just wanted to keep the HTML as clean as possible, so `cloneElement` with `hidden: true` is just a slightly esoteric alternative to 

```
<div className="hidden">
   <div className="search-results__pages">
      // etc
  </div>
</div>
```

File: site/search/SearchPanel.tsx
Comment (ikesau) [reply site/search/SearchPanel.tsx:475]: yup, it was just adding something pointless like `&[facet]=None` all the time without this.

**Comment by ikesau** [comment]
Thank you @mlbrgl ! I've made updates on all your comments (except for the Autocomplete height thing, which I wasn't able to get around to)

I think @marcelgerber might like to double-check `indexExplorersToAlgolia.ts` , and otherwise will leave this PR in your hands for the next 2 weeks. Feel free to make any further changes/merge it as you like. 🙏

File: baker/algolia/configureAlgolia.ts
Comment (marcelgerber) [inline baker/algolia/configureAlgolia.ts:123]: Both of these don't exist on explorers:

```suggestion
        customRanking: ["desc(vies_7d)"],
```

File: baker/algolia/configureAlgolia.ts
Comment (marcelgerber) [inline baker/algolia/configureAlgolia.ts:124]:
```suggestion
        attributesToSnippet: ["subtitle:24"],
```

File: site/search/SearchPanel.tsx
Comment (marcelgerber) [inline site/search/SearchPanel.tsx:287]: Hm, I haven't extensively checked this but there's definitely [a builtin `createInsightsMiddleware`](https://www.algolia.com/doc/api-reference/widgets/insights/js/) in instantsearch, maybe it's better to use that?

**Review by marcelgerber — COMMENTED** [review]
Haven't looked at all the things, I'm just bringing up a few things that jumped out. (but let me know if you want another set of eyes over something @mlbrgl!)

File: baker/algolia/indexExplorersToAlgolia.ts
Comment (mlbrgl) [reply baker/algolia/indexExplorersToAlgolia.ts:61]: Yep that doesn't make sense to me. As far as I can tell, there isn't any at the moment.

There is an odd explorer with a top-top (as in not top of the `blocks` array, but top of the config object) level "type": "Linechart" but this seems like an outdated (unpublished) format, e.g. http://datasette-private/owid?sql=select+*+from+explorers+where+config-%3E%3E%27%24.type%27+IS+NOT+NULL+

Comment (marcelgerber) [reply baker/algolia/indexExplorersToAlgolia.ts:61]: Ah yes, I think that was Breck's venture into tsv-based grapher configs.

File: baker/algolia/configureAlgolia.ts
Comment (mlbrgl) [reply baker/algolia/configureAlgolia.ts:124]: @marcelgerber [Ike seemed to be thinking that we shouldn't show subtitles for now](https://github.com/owid/owid-grapher/blob/e00f3f619e10a3b6aefcbaa1ec5006460de2d914/site/search/SearchPanel.tsx#L125). Let's revisit later?

Comment (marcelgerber) [reply baker/algolia/configureAlgolia.ts]: Makes sense! Doesn't really matter whether we already include it (and don't use it) or drop it entirely 👍🏻

**Comment by mlbrgl** [comment]
@marcelgerber thanks for the review! No need to go further, explorer indexing was the main thing that needed another set of eyes at this point.

**Comment by mlbrgl** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2533** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2533" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2533?utm_source=stack-comment).

**Pull request merged**
